### PR TITLE
[tokenbank] implement transfer-related modules

### PR DESCRIPTION
**Pull request by dong77**

### CODE REVIEW

File: packages/tokenbank_v1/contracts/modules/security/RecoveryModule.sol [hunk 1]
```diff
@@ -144,30 +149,26 @@ contract RecoveryModule is SecurityModule
...
+        require (
+            method == this.startRecovery.selector ||
+            method == this.cancelRecovery.selector ||
+            method == this.completeRecovery.selector,
+            "INVALID_METHOD"
         );
 
-        uint numSigners = (data.length - skipBytes) / 32;
+        // ASSUMPTION:
+        // data layout: {data_length:32}{wallet:32}{signers_length:32}{signer1:32}{signer2:32}
```
Comment (Brechtpd) [inline packages/tokenbank_v1/contracts/modules/security/RecoveryModule.sol]: Arrays are ABI encoded slightly differently with an offset to the data:
- https://solidity.readthedocs.io/en/v0.5.13/abi-spec.html#examples

So I think a call to e.g. `completeRecovery` needs data like this:
```
{wallet:32}{signers_offset:32}{signers_length:32}{signer1:32}{signer2:32}
```

File: packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol [hunk 2]
```diff
@@ -16,11 +16,259 @@
...
+            return;
+        }
+
+        uint allowance = ERC20(token).allowance(wallet, to);
+        if (allowance >= amount) {
+            approveInternal(wallet, token, to, amount);
+            return;
+        }
+
+        allowance -= amount;
```
Comment (Brechtpd) [inline packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol]: I think this needs to be `amount -= allowance;` to have the delta increased value to use in the following calculation.

File: packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol [hunk 3]
```diff
@@ -16,11 +16,259 @@
...
+        uint valueInCNY = priceProvider.getValueInCNY(token, allowance);
+
+        if (quotaStore.checkAndAddToSpent(wallet, valueInCNY)) {
+            approveInternal(wallet, token, to, allowance);
```
Comment (Brechtpd) [inline packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol]: Needs to be `approveInternal(wallet, token, to, amount);`?

File: packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol [hunk 4]
```diff
@@ -16,11 +16,259 @@
...
+        address wallet,
+        bytes32 pendingTxId
+        )
+        private
+        returns (bool foundPendingTx)
+    {
+        if (msg.sender != Wallet(wallet).owner()) {
+            if (isPendingTxValid(wallet, pendingTxId)) {
+                foundPendingTx = true;
+                emit PendingTxExecuted(wallet, pendingTxId, now);
```
Comment (Brechtpd) [inline packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol]: Pending transactions aren't marked as executed, so it's valid that pending transactions can be executed an unlimited amount of times within `pendingExpiry` seconds of creating the pending transaction?

File: packages/tokenbank_v1/contracts/modules/security/RecoveryModule.sol (hunk 1 above)
Comment (dong77) [reply packages/tokenbank_v1/contracts/modules/security/RecoveryModule.sol]: This is really helpful. Please take another look.

File: packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol (hunk 3 above)
Comment (dong77) [reply packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol]: it is `allowance` but it needs to be updated.

File: packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol (hunk 4 above)
Comment (dong77) [reply packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol]: fixed.

File: packages/tokenbank_v1/contracts/modules/loopring/ExchangeModule.sol [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        internal
+        view
+        returns (address[] memory signers)
+    {
+        require (
+            method == this.createDEXAccount.selector,
+            "INVALID_METHOD"
+        );
+        // ASSUMPTION:
+        // data layout: {data_length:32}{wallet:32}{signers_offset:32}{signers_length:32}{signer1:32}{signer2:32}
```
Comment (Brechtpd) [inline packages/tokenbank_v1/contracts/modules/loopring/ExchangeModule.sol]: Looks correct, but the function selector is still missing? I thought it was stripped off some place, but that doesn't seem to be the case.

File: packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol [hunk 6]
```diff
@@ -16,11 +16,257 @@
...
+        }
+
+        uint allowance = ERC20(token).allowance(wallet, to);
+        if (allowance >= amount) {
+            approveInternal(wallet, token, to, amount);
+            callContractInternal(wallet, to, 0, data);
+            return;
+        }
+
+        allowance -= amount;
```
Comment (Brechtpd) [inline packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol]: Still needs to be changed to `allowance = amount - allowance;` like above.

File: packages/tokenbank_v1/contracts/modules/loopring/ExchangeModule.sol (hunk 5 above)
Comment (dong77) [reply packages/tokenbank_v1/contracts/modules/loopring/ExchangeModule.sol]: sorry about that 👍

File: packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol (hunk 6 above)
Comment (dong77) [reply packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol]: fixed

**Pull request merged**